overlays/haskell.nix

```diff
@@ -653,11 +640,174 @@ final: prev: {
+              # Populate `depends`, `pre-existing` and `build-tools`
+              lookupDependencies = hsPkgs: depends: exe-depends: {
+                depends = final.lib.concatMap (lookupDependency hsPkgs) depends;
+                pre-existing = lookupPreExisting depends;
+                build-tools = map (lookupExeDependency hsPkgs) exe-depends;
               };
+              # Calculate the packages for a component
+              getComponents = cabal2nixComponents: hsPkgs: p:
+                let
+                  components = p.components or { ${p.component-name or "lib"} = { inherit (p) depends; exe-depends = p.exe-depends or []; }; };
+                  # Other than the `lib` and `setup` components, component names
+                  # have a prefix based on their type.
+                  componentsWithPrefix = collectionName: prefix:
+                    final.lib.listToAttrs (final.lib.concatLists (final.lib.mapAttrsToList (n: c:
+                      final.lib.optional (final.lib.hasPrefix "${prefix}:" n) (
+                        let
+                          name = final.lib.removePrefix "${prefix}:" n;
+                          value = (if cabal2nixComponents == null then {} else cabal2nixComponents.${collectionName}.${name}) // {
+                            buildable = true;
+                          } // lookupDependencies hsPkgs c.depends c.exe-depends;
+                        in { inherit name value; }
+                      )) components));
+                in
+                  final.lib.mapAttrs componentsWithPrefix haskellLib.componentPrefix
+                  // final.lib.optionalAttrs (components ? lib) {
+                    library = (if cabal2nixComponents == null then {} else cabal2nixComponents.library) // {
+                      buildable = true;
+                    } // lookupDependencies hsPkgs components.lib.depends components.lib.exe-depends;
+                  } // final.lib.optionalAttrs (components ? setup) {
+                    setup = {
+                      buildable = true;
+                    } // lookupDependencies hsPkgs.pkgsBuildBuild (components.setup.depends or []) (components.setup.exe-depends or []);
+                  };
+              callProjectResults = callCabalProjectToNix config;
+              nixFilesDir = callProjectResults.projectNix + callProjectResults.src.origSubDir or "";
+              plan-pkgs = if !builtins.pathExists (callProjectResults.projectNix + "/plan.json")
+                then
+                  # If there is no `plan.json` file assume this is a materialized
+                  # `plan-nix` and use the old code path.
+                  # TODO remove this once all the materialized files are updated
+                  importAndFilterProject {
+                    inherit (callProjectResults) projectNix sourceRepos src;
+                  }
+                else {
+                  # This replaces the `plan-nix/default.nix`
```

Comment:
This branch should be split into a separate function, like `importAndFilterProject`.